### DISCUSSION TITLE
v0.124.2 Migration Guide

### DISCUSSION THREAD

**Discussion by endigo9740** [skeletonlabs/skeleton#929] [Announcements]
As part of the [v0.124.2 release notes](https://github.com/skeletonlabs/skeleton/discussions/928) we've identified a number of breaking issues as we move to finalize many features within Skeleton. Given this, we've prepared this document to help make the transition to the new release as seamless as possible.

## Component Overhauls

As part of our component overhaul we've introduced around 16 breaking issues, with 6 components completely recreated from the ground up. However, we've tracked the breaking issues per component. View the link below and scan for items marked `BREAKING:`:

- https://github.com/skeletonlabs/skeleton/issues/857

There's too many changes for us to provide migration guides for everything, but we'll walk through a couple quick examples together.

### Minor Breaking Change

By scanning the document linked above, or viewing the [Accordion docs](https://dev.skeleton.dev/components/accordions), we'll see that `AccordionGroup` has been renamed `Accordion`. As well as the previously default auto-collapse feature now requires an `autocollapse` property. So let's implement that now: 

```html
<!-- Before -->
<AccordionGroup>...</AccordionGroup>

<!-- After -->
<Accordion autocollapse>...</Accordion>
```

### Major Refactor

Let's use [Tabs](https://www.skeleton.dev/components/tabs) as an example. Here's the before:

```ts
import { writable, type Writable } from 'svelte/store';
const storeTab: Writable<string> = writable('a');
```

```html
<TabGroup selected={storeTab}>
    <Tab value="a">(labelA)</Tab>
    <Tab value="b">(labelB)</Tab>
</TabGroup>
{#if $storeTab === 'a'}Content A{/if}
{#if $storeTab === 'b'}Content B{/if}
```

Referencing the updated [Tab documentation](https://www.skeleton.dev/components/tabs), we can see a few changes have been made:

1. The `Writable` imports are no longer needed - those can be removed
2. We now set our initial state via a simple `let` variable
3. Individual tabs now use `bind:group` and `name` props
4. A new semantic tab `panel` slot has been added to embed our tab content

Here's the changes implemented:

```ts
// NOTE: Writable import was removed

// Set our initial selection via a `let` variable:
let tabSet: string = 'a';
```
```html
<TabGroup>
    <!-- Update to use `bind:group` and `name` props -->
    <Tab bind:group={tabSet} name="a" value="a">(labelA)</Tab>
    <Tab bind:group={tabSet} name="b" value="b">(labelB)</Tab>
    <!-- Add the tab panel slot and move our condition logic within it -->
    <svelte:fragment slot="panel">
        {#if tabSet === 'a'}
            (tab panel 1 contents)
        {:else if tabSet === 'b'}
            (tab panel 2 contents)
        {/if}
    </svelte:fragment>
</TabGroup>
```

## Form Components

As part of our component overhaul updates above, we've made a series of changes to form components to operate more like native inputs and provide better support for SvelteKit native form processing. Given the scope of these changes, we've generate a detailed tutorial blog post covering this. Use this as your reference for migration:

- [Blog Post: Skeleton and Forms](https://www.skeleton.dev/blog/component-form-handling)

These changes affect the following components:

- [File Buttons](https://www.skeleton.dev/components/file-buttons) - now act as `input[type='file']` elements.
- [File Dropzones](https://www.skeleton.dev/components/file-dropzone) - now act as `input[type='file']` elements.
- [Input Chips](https://www.skeleton.dev/components/input-chips) -now  act as `select` elements.
- [Listboxes](https://www.skeleton.dev/components/listboxes) - act as `input[type='checkbox']` or `input[type='radio']`
- [Radio Groups](https://www.skeleton.dev/components/radio-groups) - now act as `input[type='radio']` elements.
- [Range Sliders](https://www.skeleton.dev/components/range-sliders) - now act as `input[type='range']` elements.
- [Slide Toggles](https://www.skeleton.dev/components/slide-toggles) - now act as `input[type='checkbox']` elements

Note that `name` is now a required property for these components. This is required to facilitate form submission.

## Action Overhauls

Thankfully these updates were fairly minor, with only one breaking change introduced for the Filters action - you're now required to pass the `#` as part of the filter ID reference. See the updated [Filter documentation](https://www.skeleton.dev/actions/filters) for details.

```html
<!-- add the # to the filter action value -->
<img src={myImageSrc} use:filter={'#BlueNight'}>
```

## Opt-In Form Styles

Form styles are no longer globally on by default, but rather opt in via the use of special classes. See the [Forms documentation](https://dev.skeleton.dev/elements/forms) for details, which is now available under Tailwind > Elements in our docs. In most cases you'll only need to implement a single class, as shown:

```html
<input class="input" ...>
<textarea class="textarea" ... />
<select class="select">...</select>
<input class="checkbox" ... />
<input class="radio" ... />
```

The `.input` styles will actually automatically handle a variety of different input types for you, including `file` and `color` type inputs. But given the difference in display for textarea, select, checkbox, and radio inputs, we've kept those classes unique.

Additionally, we recommend viewing the documentation to learn about the new validation and variant style classes that are now available. Similar badges, cards, etc, these classes are chained in alongside your primary class:

```html
<input ... class="input variant-form-material input-success" />
```

## Data Tables Delisted

As part of this release, we're delisting the Data Table utility feature and are no longer providing support for this feature. We've never recommended this feature for production use, but if you are using it in production, we recommend migrating to an alternative asap. Here's a short list of alternatives suggested by community members:

- https://svelte-headless-table.bryanmylee.com/#headless
- https://tabulator.info/
- https://tanstack.com/

> NOTE: Skeleton does not currently have an officially recommended alternative, but we do hope to add one in the future.

# Additional Support

If you need further assistance feel free to reach out in the comments section below, create a new support ticket here on GitHub, or reach out in the `#support` channel on our Discord.

https://discord.gg/EXqV7W8MtY

**Comment by knd775** [comment]
Ah, my favorite number "a" 😄

> ```ts
> // Set our initial selection via a `let` variable:
> let tabSet: number = 'a';
> ```

**Reply by endigo9740** [reply]
Doh! Fixed :D

**Comment by AgarwalPragy** [comment]
What are the benefits of using `input input-success` over the older style of using `input variant-filled-success`?

The older style still works as expected, but yields slightly different results.

`input input-success`
<img width="398" alt="image" src="https://user-images.githubusercontent.com/7423639/217649524-d8cd933b-bb8e-440b-b048-5573eb67e1b7.png">

`input variant-filled-success` 
<img width="404" alt="image" src="https://user-images.githubusercontent.com/7423639/217649477-973e5a81-be3f-4b63-9d4e-c36aa4ad3915.png">

**Reply by endigo9740** [reply]
@AgarwalPragy we may need better info in the docs for this, but the intention is for the validation state class to be added to the wrapping group element, not the individual inputs. I suppose there's use cases for both. But by targeting the group it will even work with the material variant.

The biggest benefit is the addition of the `placeholder` styles. Those are not a normal part of `.variant-filled-success`. I'm hedging my bets here that we may need to expand with other input-specific style adjustments, so I'm keen to keep these unique and separated for that reason. 

Unfortunately the placeholder styles are vendor specific, so we have to have three instances per each set. Super fun stuff :/
https://github.com/skeletonlabs/skeleton/blob/dev/src/lib/styles/elements/forms.css#L215
